runtime/goroutine.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gxruntime
+
+import (
+	"fmt"
+	"os"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ignoreRecover = false
+)
+
+// GoSafely wraps a `go func()` with recover()
+func GoSafely(wg *sync.WaitGroup, handler func(), finalFunc func(r interface{})) {
+	if wg != nil {
+		wg.Add(1)
+	}
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				if !ignoreRecover {
+					fmt.Fprintf(os.Stderr, "%s goroutine panic: %v\n%s\n",
+						time.Now(), r, string(debug.Stack()))
+				}
+				if finalFunc != nil {
+					if wg != nil {
+						wg.Add(1)
+					}
+					go func() {
+						defer func() {
+							if p := recover(); p != nil {
+								if !ignoreRecover {
+									fmt.Fprintf(os.Stderr, "recover goroutine panic:%v\n%s\n",
+										p, string(debug.Stack()))
+								}
+							}
+
+							if wg != nil {
+								wg.Done()
+							}
+						}()
+						finalFunc(r)
+					}()
+				}
+			}
+			if wg != nil {
+				wg.Done()
+			}
+		}()
+		handler()
+	}()
+}
+
+func GoUnterminal(wg *sync.WaitGroup, handle func()) {
+	GoSafely(wg,
```

Comment:
What will happen if `handler` always panic? I think the execution will be endless. handler panic -> final function -> handler panic -> final function ...

Reply:
yes, as its name said, this goroutine will run endless. This func is used for which goroutine wanna long live as its process.